File: backend/app/api/routes/reviews.py
from datetime import datetime
from typing import Any

from fastapi import APIRouter, File, Form, Path, Query, UploadFile

from app.api.deps import CurrentUser, SessionDep
from app.api.responses import csv_response
from app.csv_import import attach_import_metadata, parse_import_upload
from app.enums import MediaType
from app.schemas.response import ApiResponse, PagedApiResponse, api_page, api_success
from app.schemas.review import ReviewCreate, ReviewRead, ReviewUpdate
from app.services import permission_service, review_service, tabular_import_service

router = APIRouter(prefix="/reviews", tags=["评审 / reviews"])
router_views = APIRouter(tags=["评审 / reviews"])


@router.post("/imports", summary="导入评审 / Import Reviews")
async def import_reviews(
    session: SessionDep,
    current_user: CurrentUser,
    project_id: int = Form(...),
    collection_id: int = Form(...),
    file: UploadFile = File(...),
    dry_run: bool = Form(True),
) -> Any:
    """校验或原子导入评审。 / Validate or atomically import reviews."""
    permission_service.require_collection_resource_permission(
        session,
        collection_id=collection_id,
        project_id=project_id,
        user=current_user,
        resource_type="review",
        action="write",
        denied_detail="No review:write permission on collection",
    )
    parsed = parse_import_upload(file.filename or "", await file.read())
    report = tabular_import_service.import_reviews(
        session,
        parsed.text,
        current_user,
        project_id,
        collection_id,
        dry_run=dry_run,
    )
    return api_success(message="Import validation completed" if dry_run else "Import completed", data=attach_import_metadata(report, parsed, dry_run=dry_run))


@router.get("", response_model=PagedApiResponse[list[ReviewRead]], summary="获取评审列表 | Get reviews list")
def read_reviews(
    session: SessionDep,
    current_user: CurrentUser,
    project_id: int | None = Query(None, description="项目 ID 筛选 | Filter by project ID"),
    collection_id: int | None = Query(None, description="集合 ID 筛选 | Filter by collection ID"),
    page: int = Query(1, ge=1, description="页码 | Page number"),
    page_size: int = Query(20, ge=1, le=100, description="每页数量 | Items per page"),
    annotation_id: int | None = Query(None, description="注释 ID 筛选 | Filter by annotation ID"),
    media_name: str | None = Query(None, description="媒体名称筛选 | Filter by media name"),
    media_type: MediaType | None = Query(None, description="媒体类型精确筛选 | Exact media type filter"),
    reviewer_id: int | None = Query(None, description="评审者 ID 筛选 | Filter by reviewer ID"),
    reviewer_name: str | None = Query(None, description="评审者名称模糊筛选（大小写不敏感） | Fuzzy filter by reviewer name (case-insensitive)"),
    status_id: int | None = Query(None, description="状态 ID 筛选 | Filter by status ID"),
    status_name: str | None = Query(None, description="状态名称模糊筛选（大小写不敏感） | Fuzzy filter by status name (case-insensitive)"),
    taxon_id: int | None = Query(None, description="物种 ID 筛选 | Filter by taxon ID"),
    taxon_name: str | None = Query(None, description="物种名称模糊筛选（大小写不敏感） | Fuzzy filter by taxon name (case-insensitive)"),
    note: str | None = Query(None, description="备注筛选 | Filter by note content"),
    creation_date_from: datetime | None = Query(None, description="创建时间起 | Creation date from"),
    creation_date_to: datetime | None = Query(None, description="创建时间止 | Creation date to"),
    order_by:  str = Query("creation_date", description="排序字段 | Sort by field"),
    order_dir: str = Query("desc", description="排序方向 asc/desc | Sort direction"),
) -> Any:
    """
    获取满足条件的评审列表，按权限过滤数据。
    
    Get a paginated list of reviews matching the criteria, masked by read permissions.
    """
    filters = {
        k: v for k, v in {
            "annotation_id": annotation_id,
            "media_name": media_name,
            "media_type": media_type,
            "reviewer_id": reviewer_id,
            "reviewer_name": reviewer_name,
            "status_id": status_id,
            "status_name": status_name,
            "taxon_id": taxon_id,
            "taxon_name": taxon_name,
            "note": note,
            "creation_date_from": creation_date_from,
            "creation_date_to": creation_date_to,
            "project_id": project_id,
            "collection_id": collection_id,
        }.items() if v is not None
    }

    items, total = review_service.list_reviews(
        session=session,
        user=current_user,
        page=page,
        page_size=page_size,
        order_by=order_by,
        order_dir=order_dir,
        **filters
    )
    
    return api_page(data=items, total=total, page=page, page_size=page_size)


@router.get("/exports", summary="导出评审数据 | Export reviews data")
def export_reviews(
    session: SessionDep,
    current_user: CurrentUser,
    project_id: int | None = Query(None, description="项目 ID 筛选 | Filter by project ID"),
    collection_id: int | None = Query(None, description="集合 ID 筛选 | Filter by collection ID"),
    annotation_id: int | None = Query(None, description="注释 ID 筛选 | Filter by annotation ID"),
    media_name: str | None = Query(None, description="媒体名称筛选 | Filter by media name"),
    reviewer_id: int | None = Query(None, description="评审者 ID 筛选 | Filter by reviewer ID"),
    status_id: int | None = Query(None, description="状态 ID 筛选 | Filter by status ID"),
    taxon_id: int | None = Query(None, description="物种 ID 筛选 | Filter by taxon ID"),
    note: str | None = Query(None, description="备注筛选 | Filter by note content"),
    creation_date_from: datetime | None = Query(None, description="创建时间起 | Creation date from"),
    creation_date_to: datetime | None = Query(None, description="创建时间止 | Creation date to"),
    order_by:  str = Query("creation_date", description="排序字段 | Sort by field"),
    order_dir: str = Query("desc", description="排序方向 asc/desc | Sort direction"),
):
    """
    导出满足条件的评审数据为 CSV 格式文件。
    
    Export matching reviews to a CSV file.
    """
    filters = {
        k: v for k, v in {
            "project_id": project_id,
            "collection_id": collection_id,
        }.items() if v is not None
    }

    csv_content = review_service.export_review_csv(
        session=session,
        user=current_user,
        order_by=order_by,
        order_dir=order_dir,
        **filters,
    )
    
    return csv_response(csv_content, "reviews.csv")


@router.post(
    "",
    response_model=ApiResponse[None],
    status_code=201,
    summary="创建评审 | Create review",
)
def create_review(
    session: SessionDep,
    current_user: CurrentUser,
    data: ReviewCreate,
) -> Any:
    """
    为指定标注创建评审记录，评审者为当前登录用户。创建成功后自动标记对应的 annotation 任务为 reviewed。

    Create a review for an annotation. The reviewer is the current user. On api_success the corresponding annotation task is marked as reviewed.
    """
    review_service.create_review(
        session=session,
        user=current_user,
        data=data,
    )
    return api_success()


@router_views.put("/annotations/{annotation_id}/reviews/{reviewer_id}", response_model=ApiResponse[None], summary="更新评审信息 | Update review information")
def update_review(
    *,
    session: SessionDep,
    current_user: CurrentUser,
    project_id: int = Query(..., description="项目 ID | Project ID"),
    annotation_id: int = Path(..., description="注释 ID | Annotation ID"),
    reviewer_id: int = Path(..., description="评审者 ID | Reviewer ID"),
    review_update: ReviewUpdate,
) -> Any:
    """
    修改评审信息，需要所在集合的 review:write 权限。
    
    Update review data, requires review:write permission on the collection.
    """
    update_data = review_update.model_dump(exclude_unset=True)
    review_service.update_review(
        session=session,
        user=current_user,
        project_id=project_id,
        annotation_id=annotation_id,
        reviewer_id=reviewer_id,
        update_data=update_data
    )
    return api_success()


@router_views.delete(
    "/annotations/{annotation_id}/reviews/{reviewer_id}",
    response_model=ApiResponse[dict],
    summary="删除评审 | Delete review",
)
def delete_review(
    *,
    session: SessionDep,
    current_user: CurrentUser,
    project_id: int = Query(..., description="项目 ID | Project ID"),
    annotation_id: int = Path(..., description="注释 ID | Annotation ID"),
    reviewer_id: int = Path(..., description="评审者 ID | Reviewer ID"),
) -> Any:
    """
    删除评审记录。删除后自动将对应的 annotation 任务状态回退为 assigned。仅限管理员、集合管理者或评审者本人操作。

    Delete a review. Reverts the corresponding annotation task to assigned. Only admins, collection managers or the reviewer can delete.
    """
    review_service.delete_review(
        session=session,
        user=current_user,
        project_id=project_id,
        annotation_id=annotation_id,
        reviewer_id=reviewer_id,
    )
    return api_success(data={"message": "Review deleted successfully"})

File: frontend/src/api/endpoints/reviews.ts
import { apiClient } from "../client"
import type { FilterOptionReviewStatus, FilterOptionTaxon, FilterOptionUser } from "../utils"
import type { RowCapabilities } from "../capabilities"

/** 与 GET 标注详情内嵌 reviews、POST/PUT 响应一致 */
export interface AnnotationReviewRead {
    annotation_id: number
    reviewer_id: number
    annotation_review_status_id: number
    taxon_id?: number | null
    note?: string | null
    creation_date: string
    media_name?: string | null
    media_type?: string | null
    reviewer_name: string
    status_name: string
    taxon_name?: string | null
    capabilities?: RowCapabilities
}

export interface ReviewCreatePayload {
    project_id: number
    annotation_id: number
    annotation_review_status_id: number
    taxon_id?: number | null
    note?: string | null
}

export interface ReviewUpdatePayload {
    annotation_review_status_id: number
    taxon_id?: number | null
    note?: string | null
}

/** GET /v1/reviews 查询参数，与当前后端路由一致。 */
export interface ReviewsListParams {
    project_id?: number | null
    collection_id?: number | null
    page?: number
    page_size?: number
    annotation_id?: number | null
    media_name?: string | null
    media_type?: string | null
    reviewer_id?: number | null
    reviewer_name?: string | null
    status_id?: number | null
    status_name?: string | null
    taxon_id?: number | null
    taxon_name?: string | null
    note?: string | null
    creation_date_from?: string | null
    creation_date_to?: string | null
    order_by?: string
    order_dir?: "asc" | "desc" | string
}

export interface ReviewsExportParams {
    project_id?: number | null
    collection_id?: number | null
    annotation_id?: number | null
    media_name?: string | null
    media_type?: string | null
    reviewer_id?: number | null
    reviewer_name?: string | null
    status_id?: number | null
    status_name?: string | null
    taxon_id?: number | null
    taxon_name?: string | null
    note?: string | null
    creation_date_from?: string | null
    creation_date_to?: string | null
    order_by?: string
    order_dir?: "asc" | "desc" | string
}

export interface ReviewsPageInfo {
    total: number
    page: number
    page_size: number
    total_pages?: number
}

type ReviewsPagedBody = {
    code: number
    message: string
    data: AnnotationReviewRead[] | null
    page_info: ReviewsPageInfo | null
}

function cleanParams(params: Record<string, unknown>): Record<string, string | number | boolean | undefined> {
    return Object.fromEntries(
        Object.entries(params).filter(([, v]) => v !== undefined && v !== null && String(v).trim() !== ""),
    ) as Record<string, string | number | boolean | undefined>
}

export const reviewsApi = {
    /** Data > Reviews：筛选下拉（project_id 必填） */
    getFilterOptions(params: { project_id: number; collection_id?: number }) {
        return apiClient.get<{
            code: number
            message: string
            data: { reviewer: FilterOptionUser[]; status: FilterOptionReviewStatus[]; taxon: FilterOptionTaxon[] }
        }>("/v1/reviews/filter-options", { params: cleanParams(params) })
    },

    /** 获取 Reviews 分页列表（原始响应） */
    getList(params: ReviewsListParams, ignoreUnauthorized?: boolean) {
        return apiClient.get<ReviewsPagedBody>("/v1/reviews", {
            params: cleanParams(params as Record<string, unknown>),
            ignoreUnauthorized,
        })
    },

    /** GET /v1/reviews — 解析 code / data / page_info */
    async listPaged(params: ReviewsListParams, ignoreUnauthorized?: boolean): Promise<{
        items: AnnotationReviewRead[]
        pageInfo: ReviewsPageInfo
    }> {
        const res = await apiClient.get<ReviewsPagedBody>("/v1/reviews", {
            params: cleanParams(params as Record<string, unknown>),
            ignoreUnauthorized,
        })
        if (res.code !== 0) {
            throw new Error(res.message || `Request failed (code ${String(res.code)})`)
        }
        const page = typeof params.page === "number" ? params.page : 1
        const pageSize = typeof params.page_size === "number" ? params.page_size : 20
        return {
            items: res.data ?? [],
            pageInfo:
                res.page_info ?? {
                    total: 0,
                    page,
                    page_size: pageSize,
                    total_pages: 0,
                },
        }
    },

    /**
     * 创建评审。后端成功时可能返回 `data: null`，需再 GET 列表刷新。
     */
    async create(payload: ReviewCreatePayload): Promise<AnnotationReviewRead | null> {
        const res = await apiClient.post<{
            code: number
            message: string
            data: AnnotationReviewRead | null
        }>("/v1/reviews", payload)
        if (res.code !== 0) {
            throw new Error(res.message || `Request failed (code ${String(res.code)})`)
        }
        return res.data ?? null
    },

    /**
     * 更新评审。后端成功时可能返回 `data: null`，需再 GET 列表刷新。
     */
    async update(
        annotationId: number,
        reviewerId: number,
        projectId: number,
        payload: ReviewUpdatePayload,
    ): Promise<AnnotationReviewRead | null> {
        const res = await apiClient.put<{
            code: number
            message: string
            data: AnnotationReviewRead | null
        }>(
            `/v1/annotations/${annotationId}/reviews/${reviewerId}`,
            payload,
            { params: { project_id: projectId } },
        )
        if (res.code !== 0) {
            throw new Error(res.message || `Request failed (code ${String(res.code)})`)
        }
        return res.data ?? null
    },

    /**
     * 删除评审。
     */
    async delete(annotationId: number, reviewerId: number, projectId: number): Promise<void> {
        const res = await apiClient.delete<{ code: number; message: string }>(
            `/v1/annotations/${annotationId}/reviews/${reviewerId}`,
            { params: { project_id: projectId } },
        )
        if (res.code !== 0) {
            throw new Error(res.message || `Request failed (code ${String(res.code)})`)
        }
    },

    exportCsv(params: ReviewsExportParams) {
        return apiClient.download("/v1/reviews/exports", {
            params: cleanParams(params as Record<string, unknown>),
        })
    },
}
